src/Avalonia.Base/Media/TextFormatting/ShapedBuffer.cs

```diff
@@ -89,90 +90,106 @@ public GlyphInfo this[int index]
 
         public IEnumerator<GlyphInfo> GetEnumerator() => _glyphInfos.GetEnumerator();
 
+        internal void ResetBidiLevel(sbyte paragraphEmbeddingLevel) => BidiLevel = paragraphEmbeddingLevel;
+
+        int IReadOnlyCollection<GlyphInfo>.Count => _glyphInfos.Length;
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
         /// <summary>
-        /// Finds a glyph index for given character index.
+        /// Splits the <see cref="TextRun"/> at specified length.
         /// </summary>
-        /// <param name="characterIndex">The character index.</param>
-        /// <returns>
-        /// The glyph index.
-        /// </returns>
-        private int FindGlyphIndex(int characterIndex)
+        /// <param name="textLength">The text length.</param>
+        /// <returns>The split result.</returns>
+        public SplitResult<ShapedBuffer> Split(int textLength)
         {
-            if (characterIndex < _glyphInfos[0].GlyphCluster)
+            // make sure we do not overshoot
+            textLength = Math.Min(Text.Length, textLength);
+
+            if (textLength <= 0)
```

Comment:
I'd argue that splitting with a length of 0 should be valid, simply returning the buffer itself (like it's done below when the split is beyond the length).

Reply:
Sorry @Gillibald, why do we need to return `(empty, this)` in this case instead of just returning `(this, null)` - i.e. not splitting at all? I'm a little worried of some unrelated loop choking on the empty buffer.

We don't return `(this, empty)` when the split is right after the buffer.

Or am I overthinking this, and this scenario (splitting at 0) never happens in reality?

Reply:
Yes, splitting at zero never happens intentionally. To detect that an exception was thrown before. The second part is always the remaining part of the split operation, for a zero length, the original buffer.